### PR TITLE
[terra-paginator] Updated uncontrolled responsive-element to controlled responsive-element

### LINKED ISSUE #2746: [terra-paginator] Uplift to the controlled responsive element
# Feature Request

Investigate uplifting terra-paginator onto the controlled responsive element or alternative. 

## Description

The uncontrolled responsive element has been [flagged for deprecation](https://github.com/cerner/terra-core/pull/2729). The terra-paginator should either be uplifting onto the controlled responsive element implementation or alternatively the responsive element should be removed from the component.

### PR DESCRIPTION
**Pull request by pranav300**
### Summary
Resolves #2746

### CODE REVIEW

File: packages/terra-paginator/src/ControlledPaginator.jsx [hunk 1]
```diff
@@ -262,7 +273,12 @@ class Paginator extends React.Component {
   }
 
   render() {
-    return <ResponsiveElement tiny={this.reducedPaginator()} small={this.defaultPaginator()} />;
+    return (
+      <ResponsiveElement onChange={value => this.setState({ breakpoint: value })}>
```
Comment (emilyrohrbough) [inline packages/terra-paginator/src/ControlledPaginator.jsx]: Should we do something like the following so we only update state when transitioning between tiny and non-tiny?
```js
<ResponsiveElement
    onChange={value => {
        const showReducedPaginator = value === 'tiny';

        if (this.state.showReducedPaginator !== showReducedPaginator) {
            this.setState({ showReducedPaginator })
        }
    }}
>
    { this.state.showReducedPaginator ? this.reducedPaginator() : this.defaultPaginator()}
</ResponsiveElement>
```

File: packages/terra-paginator/src/Paginator.jsx [hunk 2]
```diff
@@ -276,7 +284,12 @@ class Paginator extends React.Component {
   }
 
   render() {
-    return <ResponsiveElement tiny={this.reducedPaginator()} medium={this.defaultPaginator()} />;
+    return (
```
Comment (emilyrohrbough) [inline packages/terra-paginator/src/Paginator.jsx]: same comment as above: https://github.com/cerner/terra-core/pull/2751/files#r355049197

File: packages/terra-paginator/src/ProgressivePaginator.jsx [hunk 3]
```diff
@@ -211,7 +219,14 @@ class ProgressivePaginator extends React.Component {
   }
 
   render() {
-    return <ResponsiveElement tiny={this.reducedProgressivePaginator()} small={this.defaultProgressivePaginator()} />;
+    return (
+      <ResponsiveElement onChange={value => this.setState({ breakpoint: value })}>
```
Comment (emilyrohrbough) [inline packages/terra-paginator/src/ProgressivePaginator.jsx]: same question as above: https://github.com/cerner/terra-core/pull/2751/files#r355049197

File: packages/terra-paginator/tests/jest/ControlledProgressivePaginator.test.jsx [hunk 4]
```diff
@@ -7,8 +7,18 @@ describe('Paginator', () => {
...
-  it('should render a Controlled ProgressivePaginator', () => {
+  it('should render a Controlled ProgressivePaginator - Tiny', () => {
     const wrapper = shallowWithIntl(defaultRender).dive();
+    wrapper.setState({ breakpoint: 'tiny' });
+    wrapper.update();
+    expect(wrapper).toMatchSnapshot();
+  });
+
+  // Snapshot Tests
+  it('should render a Controlled ProgressivePaginator - LArge', () => {
```
Comment (emilyrohrbough) [inline packages/terra-paginator/tests/jest/ControlledProgressivePaginator.test.jsx]:
```suggestion
  it('should render a Controlled ProgressivePaginator - Large', () => {
```

**Comment by nramamurth** [comment]
@emilyrohrbough 

> Why did the wdio screenshots update? Is that expected?

I remember @pranav300 mentioning to me that this is due to an existing bug in the snippet below ⬇️ `defaultElement` prop was removed from terra-responsive-element@v5.0.0 (a long time ago) but looks like it's still hanging out here. 

https://github.com/cerner/terra-core/blob/bdd0dba157ba013f72493e9e5476fa54cfb3deb4/packages/terra-paginator/src/ControlledProgressivePaginator.jsx#L201-L203

File: packages/terra-paginator/tests/jest/ControlledProgressivePaginator.test.jsx (hunk 4 above)
Comment (pranav300) [reply packages/terra-paginator/tests/jest/ControlledProgressivePaginator.test.jsx]: Fixed here 93f9aa9.

File: packages/terra-paginator/src/ProgressivePaginator.jsx (hunk 3 above)
Comment (pranav300) [reply packages/terra-paginator/src/ProgressivePaginator.jsx]: Fixed here 93f9aa9.

File: packages/terra-paginator/src/Paginator.jsx (hunk 2 above)
Comment (pranav300) [reply packages/terra-paginator/src/Paginator.jsx]: Fixed here 93f9aa9.

File: packages/terra-paginator/src/ControlledPaginator.jsx (hunk 1 above)
Comment (pranav300) [reply packages/terra-paginator/src/ControlledPaginator.jsx]: Fixed here 93f9aa9.

File: packages/terra-paginator/src/ProgressivePaginator.jsx [hunk 5]
```diff
@@ -42,6 +42,7 @@ class ProgressivePaginator extends React.Component {
 
     this.state = {
       selectedPage: selectedPage && totalCount ? selectedPage : undefined,
+      showReducedPaginator: undefined,
```
Comment (ShettyAkarsh) [inline packages/terra-paginator/src/ProgressivePaginator.jsx]: should this value be false ?

File: packages/terra-paginator/src/ControlledPaginator.jsx [hunk 6]
```diff
@@ -262,7 +265,19 @@ class Paginator extends React.Component {
   }
 
   render() {
-    return <ResponsiveElement tiny={this.reducedPaginator()} small={this.defaultPaginator()} />;
+    return (
+      <ResponsiveElement
+        onChange={value => {
```
Comment (ShettyAkarsh) [inline packages/terra-paginator/src/ControlledPaginator.jsx]: Just for readability. you could add this logic in separate event.

File: packages/terra-paginator/src/ProgressivePaginator.jsx [hunk 7]
```diff
@@ -211,7 +219,15 @@ class ProgressivePaginator extends React.Component {
   }
 
   render() {
-    return <ResponsiveElement tiny={this.reducedProgressivePaginator()} small={this.defaultProgressivePaginator()} />;
+    return (
+      <ResponsiveElement
+        onChange={value => {
```
Comment (emilyrohrbough) [inline packages/terra-paginator/src/ProgressivePaginator.jsx]: couldn't this be `onChange{this.setPaginator}`

File: packages/terra-paginator/src/ControlledPaginator.jsx [hunk 8]
```diff
@@ -46,6 +46,17 @@ class Paginator extends React.Component {
...
     this.reducedPaginator = this.reducedPaginator.bind(this);
+    this.setPaginator = this.setPaginator.bind(this);
+    this.state = {
+      showReducedPaginator: false,
+    };
+  }
+
+  setPaginator(value) {
+    const showReducedPaginator = value === 'tiny';
+    if (this.state.showReducedPaginator !== showReducedPaginator) {
```
Comment (nramamurth) [inline packages/terra-paginator/src/ControlledPaginator.jsx]: I'd expect this to be - 

```
this.setState((previousState) => 
    { showReducedPaginator:  previousState.showReducedPaginator !== showReducedPaginator}
);
```

File: packages/terra-paginator/CHANGELOG.md [hunk 9]
```diff
@@ -4,6 +4,9 @@ ChangeLog
 Unreleased
 ----------
 
+### Changed
+* Updated `uncontrolled responsive-element` to `controlled responsive-element`.
```
Comment (nramamurth) [inline packages/terra-paginator/CHANGELOG.md]:
```suggestion
* Replaced uncontrolled `<ResponsiveElement>` with controlled `<ResponsiveElement>`.
```

File: packages/terra-overlay/CHANGELOG.md [hunk 10]
```diff
@@ -6,6 +6,9 @@ Unreleased
 
 3.38.0 - (January 7, 2020)
 ------------------
+### Changed
+* Updated screenshot
```
Comment (StephenEsser) [inline packages/terra-overlay/CHANGELOG.md]: This entry is in a released section. Was the updated screenshot diff removed?

Comment (emilyrohrbough) [reply packages/terra-overlay/CHANGELOG.md]: this was cleaned up in this pr: https://github.com/cerner/terra-core/pull/2791

it was a change that resulted from reverting the axe-xore upgrade in toolki.

**Pull request merged**